module#2/src/api/controllers/users-controllers.ts

```diff
@@ -0,0 +1,83 @@
+import { Request, Response } from "express";
+import { dataBase, ERROR_MESSAGE, HTTP_STATUSES } from "../../constants";
+import { v4 as id } from "uuid";
+import { getAutoSuggestUsers, isUserAlreadyExist } from "../../utils/user-operations";
+
+export const getAllUsers = (req: Request, res: Response) => {
+  const { limit, loginSubstring } = req.query;
+
+  if (limit || loginSubstring) {
+    const findUsers = getAutoSuggestUsers(String(loginSubstring), Number(limit));
+
+    return res.json(findUsers);
+  }
+
+  return res.json(dataBase.users);
+};
+
+export const getUserById = (req: Request, res: Response) => {
+  const foundUser = dataBase.users.find((user) => user.id === req.params.userId);
+
+  if (!foundUser) {
+    res.sendStatus(HTTP_STATUSES.NOT_FOUND_404);
+    return;
+  }
+
+  res.json(foundUser);
+};
+
+export const updateUserById = (req: Request, res: Response) => {
+  const indexFoundUser = dataBase.users.findIndex((user) => user.id === req.params.userId);
+
+  if (!indexFoundUser) {
+    res.sendStatus(HTTP_STATUSES.NOT_FOUND_404);
+    return;
+  }
+
+  const user = req.body;
+
+  if (isUserAlreadyExist(user.login)) {
+    return res.status(HTTP_STATUSES.BAD_REQUEST_400).json(ERROR_MESSAGE.USER_ALREADY_EXIST);
+  }
```

Comment:
I guess this is a a part of validator logic, shouldn't be here. I suggest create separate validator, attach it as a middleware. And also, validator for create will be different from update. Please fix this also.

Reply:
Yes, thanks for your remark) Fixed, however there is one question. Why do you think that validator for create operation and updated has to be different??? here we just check out existing login name. The logic the same for both cases. If I am wrong, please correct me.

Reply:
Since we are building restful web service, there always should be difference between create: i.e. post VS. update: i.e. put/patch, but with an update we never update id: uuid, or in some table schemas login: uuid. So, for create we have to verify that login and id is send, but for update they can be optional. Because one of them is not changeable because it is unique by default, our service shouldn't be able to update uuid, otherwise we will loose reference to the row with this id. Besides our table is in memory, and if we have more than one object with same id, then it will require extra logic to distinguish what data we want to get.

Reply:
Thank you very much for such describing explanation) And I also thought that `update` has to allow change not every field. And I believe it also can depend on requirements, I mean which field customer, business logic allow change, and which nor in different circumstances))))